File: python/stream_decoder.py
from dataclasses import dataclass, field, fields
from typing import Optional
import json
from collections import namedtuple

@dataclass
class LLMChunk:
    content: str
    sentTime: Optional[int] = field(default=None)
    isFinal: Optional[bool] = field(default=None)

@dataclass
class LLMReply:
    requestId: int
    userInput: str
    success: bool
    ttftMsec: int
    ttltMsec: int
    tokenCount: int
    wordCount: int
    errMsg: str
    isFinal: Optional[bool] = field(default=None)

class StreamDecoder():
    type_map = {
        'LLMChunk': LLMChunk,
        'LLMReply': LLMReply
    }

    def __init__(self):
        self.partial_undecoded = bytearray()

    def decode(self, stream_chunk):
        self.partial_undecoded += stream_chunk
        try:
            undecoded_chunks = self.partial_undecoded.decode('utf-8')
        except UnicodeDecodeError:
            return []

        decoded_chunks = []
        if undecoded_chunks.endswith('\n'):
            decoded_chunks = StreamDecoder.__decode_chunks(undecoded_chunks)
            self.partial_undecoded = bytearray()

        return decoded_chunks

    @staticmethod
    def __decode_chunks(undecoded_chunks):
        undecoded_chunk_list = undecoded_chunks.split('\n')
        decoded_chunks = []
        for undecoded in undecoded_chunk_list[:-1]:
            if undecoded:
                decoded_chunks.append(StreamDecoder.__decode_chunk(undecoded))

        return decoded_chunks

    @staticmethod
    def __decode_chunk(undecoded):
        data = json.loads(undecoded)
        type_name = data.pop('typeName')

        try:
            Type = StreamDecoder.type_map[type_name]
        except KeyError:
            raise ValueError(f"Unknown type: {type_name}")

        filtered = {f.name: data[f.name] for f in fields(Type) if f.name in data}
        return Type(**filtered)

File: typescript/StreamDecoder.ts
export class LLMChunk {
    content?: string;
    sentTime?: number;
    isFinal?: boolean;

    constructor(data: Partial<LLMChunk>) {
        this.content = data.content;
        this.sentTime = data.sentTime;
        this.isFinal = data.isFinal;
    }
}

export class LLMReply {
    requestId?: number;
    userInput?: string;
    success?: boolean;
    ttftMsec?: number;
    ttltMsec?: number;
    tokenCount?: number;
    wordCount?: number;
    errMsg?: string;
    isFinal?: boolean;

    constructor(data: Partial<LLMReply>) {
        this.requestId = data.requestId;
        this.userInput = data.userInput;
        this.success = data.success;
        this.ttftMsec = data.ttftMsec;
        this.ttltMsec = data.ttltMsec;
        this.tokenCount = data.tokenCount;
        this.wordCount = data.wordCount;
        this.errMsg = data.errMsg;
        this.isFinal = data.isFinal;
    }
}

export type ValidType = LLMChunk | LLMReply;

interface Constructor<T> {
    new (data: Partial<T>): T;
}

export class StreamDecoder {
    private decoder: any;
    private partialUndecoded: string;

    public constructor() {
        this.decoder = new TextDecoder;
        this.partialUndecoded = '';
    }

    private static readonly typeMap = new Map<string, Constructor<ValidType>>([
        ['LLMChunk', LLMChunk],
        ['LLMReply', LLMReply]
    ]);

    public decode(streamChunk: Buffer): ValidType[] {
        this.partialUndecoded += this.decoder.decode(streamChunk, {stream: true});

        let decodedChunks: ValidType[] = [];
        if (this.partialUndecoded.endsWith('\n')) {
            decodedChunks = StreamDecoder.decodeChunks(this.partialUndecoded);
            this.partialUndecoded = '';
        }
        return decodedChunks;
    }

    private static decodeChunks(undecodedChunks: string): ValidType[] {
        const undecodedChunkList = undecodedChunks.split('\n');
        const nSplits = undecodedChunkList.length;

        const decodedChunks: ValidType[] = [];
        for (let i = 0; i < nSplits - 1; i++) {
            const undecoded = undecodedChunkList[i];
            if (undecoded) {
                decodedChunks.push(StreamDecoder.decodeChunk(undecoded));
            }
        }
        return decodedChunks;
    }

    private static decodeChunk(undecoded: string): ValidType {
        const data = JSON.parse(undecoded);
        const typeName = data['typeName'];
        const Type = StreamDecoder.typeMap.get(typeName);

        if (!Type) {
            throw new Error(`Unknown type: ${typeName}`);
        }

        return new Type(data);
    }
}
